README.md
# rethinker [![CRAN downloads](http://cranlogs.r-pkg.org/badges/rethinker)](https://cran.r-project.org/web/packages/rethinker/index.html)

Rethinker is a [RethinkDB](http://rethinkdb.com/) driver for [R](https://www.r-project.org/).
It is currently pretty usable, but mileage may vary; the main thing missing is the support for encryption connections, auth keys and v1.0 wire protocol.
All those things boil down to the fact that R has no support for cryptographic primitives nor for wrapping connections into TLS.

How to use
---------

The easiest way is to install from [CRAN](https://cran.r-project.org/web/packages/rethinker/index.html), i.e.,

```r
install.packages('rethinker')
```

To build from source, [see the devel branch](https://github.com/mbq/rethinker/tree/devel).

Basically, the client works like this:

```r
library(rethinker)
cn<-openConnection(host,port)

# Sync query, will just print result
r()$db("someDb")$table("someTable")$count()$run(cn)
# ... or just
r("someDb","someTable")$count()$run(cn)

# Sync query, but will return cursor...
r()$db("someDb")$table("someTable")$orderBy(list(index='awe'))$run(cn)->cr
# ...you can move with...
cursorNext(cr)
# ...or close
close(cr)

# Async query...
r()$db("someDb")$table("someTable")$changes()$runAsync(cn,
 function(x){
  print(x);
  return(FALSE); #TRUE for more
 }
)
# Because R has no event loop, this must be called for cb to be executed;
# it will block till all installed cbs will either exhaust their queries
# or kill themselves by returning FALSE.
drainConnection(cn)

# Terms are named as in JS driver, options as in Python (snake_case);
# objects are lists, mapped as in rjson package
r()$db("someDb")$table("someTable")$insert(
 list(
  id="777",
  stuff=list(who="cares")
 ),
 conflict="update",
 return_changes=TRUE
)$run(cn)

# Brackets are replaced by $bracket(), do by $funcall(function,atts)
# Implicit var ("row") is not available; R anonymous functions are quite short and easy, there is no
#  need for additional syntax sugar.
```


inst/NEWS
Changes in 1.1.0:
 - AST generation errors fixed.
 - Documentation updates.
 - Vignette with an usage tutorial.

Changes in 1.0.0:
  - First public release.


tests/testthat/testAst.R
library(rjson);

context("Abstract Syntax Tree");

test_that("Basics",{
 expect_identical(
  toJSON(r()$db("a")$table("b")$add()$query),
  "[24,[[15,[[14,[\"a\"]],\"b\"]]]]");
 expect_identical(
  toJSON(r()$db("a")$table("b")$query),
  "[15,[[14,[\"a\"]],\"b\"]]");

});

test_that("Functions",{
 Q<-r()$funcall(function(x) x,777)$query;
 expect_identical(toJSON(Q),
  "[64,[[69,[[2,[1]],[10,[1]]]],777]]")
 expect_error(r()$filter(function() r()$add(1))$query);
 expect_error(r()$filter(list(a=function(x) x))$query,
  "Functions can only exist as direct term arguments.");
})

test_that("Make array appears",{
 Q<-r()$db(c("a","b","c"))$query;
 expect_identical(toJSON(Q),"[14,[[2,[\"a\",\"b\",\"c\"]]]]");
})

test_that("Expressions",{
 Q<-r()$add(r()$add(1,2),4)$query;
 expect_identical(toJSON(Q),"[24,[[24,[1,2]],4]]");
})

test_that("Implicit var throws",{
 expect_error(r()$row("a"),"Implicit")
})


tests/testthat/testTransactions.R
context("Transactions");

require(tools)

datadir<-sprintf("%s/rethinkdb-datadir",tempdir())
logfile<-sprintf("%s/rethinkdb-log",tempdir())
pidfile<-sprintf("%s/rethinkdb-pid",tempdir())
skipAll<-FALSE
port<-'37891'
portOff<-'9876'

tryKill<-function(){
 if(file.exists(pidfile)){
  try(scan(pidfile,what=numeric(),n=1,quiet=TRUE),silent=TRUE)->pid
  if(!inherits(pid,"try-error"))
   pskill(pid)
 }
}

## Set-up a throw-away rethinkdb server
#First, is rethink instlled?
try(system2('rethinkdb','-v',stderr=TRUE,stdout=TRUE),silent=TRUE)->ver
if(inherits(ver,"try-error")){
 message("No rethinkdb installed, skipping.")
 skipAll<-TRUE
}else{
 tryKill()

 #Invoke; this will do nothing when pidfile is ok
 system2('rethinkdb',c(
  '--no-update-check',
  '--no-http-admin',
  '--port-offset',portOff,
  '--directory',datadir,
  '--cores','1',
  '--io-threads','1',
  '--daemon',
  '--log-file',logfile,
  '--pid-file',pidfile
 ),stdout=FALSE,stderr=FALSE,wait=FALSE)
 Sys.sleep(5)
 try(openConnection(port=port),silent=TRUE)->J;
 if(!inherits(J,"try-error")){
  #Init DB for testing
  r()$db("test")$tableList()$run(J)->tables;
  if("rethinker_tests"%in%tables)
  r()$db("test")$tableDrop("rethinker_tests")$run(J);
  close(J);
 }else{
  message("Running throw-away RethinkDB server failed, skipping.")
  skipAll<-TRUE
 }
}

if(!skipAll){

test_that("Connection prints",{
 J<-openConnection(port=port);
 expect_output(print(J),"Opened");
 close(J);
});

test_that("Make table",{
 J<-openConnection(port=port);
 r()$db("test")$tableCreate("rethinker_tests")$run(J)->ans;
 expect_equal(ans$tables_created,1);
 close(J);
});

test_that("Insert, read and delete an object",{
 obj<-list(
  id=0,
  a=list(b=1,c=list(ca=LETTERS,cb=letters),d=3:5),
  d=6,e="siedem",
  f=8:10);
 J<-openConnection(port=port);
 expect_equal(
  r()$db("test")$table("rethinker_tests")$insert(obj)$run(J)$inserted,
  1);
 expect_equal(
  r()$db("test")$table("rethinker_tests")$get(0)$run(J)$a$c$ca,
  LETTERS);
 expect_equal(
  r()$db("test")$table("rethinker_tests")$get(0)$delete()$run(J)$deleted,
  1);
 close(J);
});

test_that("Bulk insert, cursor",{
 lapply(1:1000,function(x) list(id=x,tester=77))->tins;
 J<-openConnection(port=port);
 expect_equal(
  r()$db("test")$table("rethinker_tests")$insert(tins)$run(J)$inserted,
  1000);
 cur<-r()$db("test")$table("rethinker_tests")$run(J);

 expect_equal(
  cursorNext(cur)$tester,
  77);
 cursorNext(cur,inBatch=TRUE)->stuff;
 expect_gt(length(stuff),10);
 expect_equal(stuff[[6]]$tester,77);
 expect_equal(
  r()$db("test")$table("rethinker_tests")$delete()$run(J)$deleted,
  1000);
 close(J);
});

test_that("Cursor emptying",{
 J<-openConnection(port=port);
 r()$db("test")$table("rethinker_tests")$changes()$run(J)->cur;
 expect_output(print(cur),"Active");
 close(cur); Sys.sleep(1);
 expect_output(print(cur),"Empty");
 expect_identical(cursorNext(cur),NULL);
 expect_identical(cursorNext(cur,inBatch=TRUE),list());

 r()$db("test")$table("rethinker_tests")$changes()$run(J)->cur;
 close(J); Sys.sleep(1);
 expect_output(print(cur),"Empty");
 expect_identical(cursorNext(cur),NULL);
 expect_identical(cursorNext(cur,inBatch=TRUE),list());
 close(cur);
});

test_that("Async queries",{
 J<-openConnection(port=port);
 had<-FALSE;
 r()$db("test")$table("rethinker_tests")$changes()$runAsync(J,function(r){
  had<<-TRUE;
  return(FALSE);
 });
 Sys.sleep(1); #Wait for changes observer to settle down
 r()$db("test")$table("rethinker_tests")$insert(list(id='zz',n=runif(10)),conflict="update")$run(J);
 drainConnection(J);
 expect_identical(had,TRUE);
 r()$db("test")$table("rethinker_tests")$delete()$run(J);
 close(J);
})

test_that("Sync-async mix",{
 J<-openConnection(port=port);
 r()$db("test")$table("rethinker_tests")$delete()$run(J);
 asyncCount<-0;
 r()$db("test")$table("rethinker_tests")$changes()$runAsync(J,function(r){
  asyncCount<<-asyncCount+1;
  return(asyncCount<5);
 });
 asyncCount2<-0;
 r()$db("test")$table("rethinker_tests")$changes()$runAsync(J,function(r){
  asyncCount2<<-asyncCount+1;
  return(asyncCount2<5);
 });
 r()$db("test")$table("rethinker_tests")$changes()$run(J)->cursor;
 Sys.sleep(1); #Wait for changes observer to settle down

 for(e in 1:5)
  r()$db("test")$table("rethinker_tests")$insert(list(id=220+e,n=runif(10)))$run(J);

 drainConnection(J);
 expect_equal(asyncCount,5);
 expect_equal(asyncCount2,5);
 cursorNext(cursor,inBatch=TRUE)->stuff;
 expect_equal(length(stuff),5);

 r()$db("test")$table("rethinker_tests")$delete()$run(J);
 close(J);
});

test_that("Profile",{
 J<-openConnection(port=port);
 expect_message(
  r()$db("test")$table("rethinker_tests")$count()$run(J,profile=TRUE),
  "Saved profile");
 expect_true(!is.null(J$lastProfile[[1]]$description));
 close(J);
});

test_that("Simple expression",{
 J<-openConnection(port=port);
 expect_equal(r()$add(r()$add(1,2),4)$run(J),7);
 close(J);
});

test_that("Map-reduce",{
 lapply(1:100,function(x) list(id=x,val=list(vall=x)))->tins;
 J<-openConnection(port=port);
 expect_equal(
  r()$db("test")$table("rethinker_tests")$insert(tins)$run(J)$inserted,
  100);

 r()$db("test")$table("rethinker_tests")$map(
  function(x) x$bracket("val")$bracket("vall")
 )$reduce(
  function(l,r) r()$add(l,r)
 )$run(J)->ans;

 expect_equal(
  r()$db("test")$table("rethinker_tests")$delete()$run(J)$deleted,
  100);
 close(J);
});

test_that("r() parameters",{
 J<-openConnection(port=port);
 expect_equal(
  r("test","rethinker_tests")$insert(list(id="xyz"))$run(J)$inserted,
  1);
 expect_equal(
  r("test","rethinker_tests")$delete()$run(J)$deleted,
  1);
 close(J);
});

test_that("Expr",{
 J<-openConnection(port=port);
 expect_equal(r()$expr(list(a=1,b=2,c=3))$keys()$run(J),c('a','b','c'));
 close(J);
});

} #Skip-all

tryKill()


R/wire.R
#' Open connection to a RethinkDB server
#'
#' Opens connection to a given RethinkDB server.
#'
#' @param host Host to connect to.
#' @param port Port to connect to.
#' @param user Username; only for \code{v} equal \code{"V1_0"}.
#' @param password Password; only for \code{v} equal \code{"V1_0"}.
#' @param v Protocol version; \code{"V0_3"} and \code{"V0_4"} supported, the default should be used.
#' @return Object of a class \code{RethinkDB_connection}, which can be passed to \code{r()$...$run} and \code{r()$...$runAsync} functions.
#' @export
openConnection<-function(host='localhost',port=28015,v="V0_4",user="admin",password=""){
 JSON<-0x7e6970c7L
 V0_3<-0x5f75e83eL
 V0_4<-0x400c2d20L
 V1_0<-0x34c2bdc3L

 #Open connection; R will throw error in case of trouble here
 socketConnection(host,port,
  open='w+b',blocking=TRUE)->con;

 if(identical(v,"V1_0")){
  if(!requireNamespace("openssl",quietly=TRUE)){
   stop("V1_0 interface requires the openssl package.")
  }

  #Send handshake
  writeBin(as.integer(V1_0),con,size=4,endian="little");

  #Pull response; either a JSON or error string
  while(TRUE){
   response<-readBin(con,character(),1);
   if(length(response)>0) break;
  }
  try(rjson::fromJSON(response),silent=TRUE)->rParsed;
  if(inherits(rParsed,"try-error")||is.null(rParsed$success)||!rParsed$success)
   stop(sprintf("Handshake ended in error: %s",response));
  if(rParsed$min_protocol_version>0)
   stop("This client is too old for your RethinkDB version.");
  haveVer<-V1_0;

  #TODO: doScrumSha256()
  stop("V1_0 is not yet supported.");
 }else{
  header<-NULL;
  if(identical(v,"V0_3")){
   header<-c(V0_3,
    as.integer(nchar(password)),
    utf8ToInt(password),
    JSON);
  }
  if(identical(v,"V0_4")){
   header<-c(V0_4,
    as.integer(nchar(password)),
    utf8ToInt(password),
    JSON);
  }
  if(is.null(header)) stop(sprintf("Unknown protocol version %s!",v));
  gotVer<-header[1];

  #Send handshake
  writeBin(as.integer(header),con,size=4,endian='little');

  #Pull response; should be, literally, "SUCCESS"
  while(TRUE){
   response<-readBin(con,character(),1);
   if(length(response)>0) break;
  }
  if(!identical(response,"SUCCESS"))
   stop(sprintf(
    "Could not establish connection with RethinkDB server.\nIt says: %s",
    response));
 }

 ans<-new.env();
 ans$con<-con;
 ans$host<-host;
 ans$port<-port;
 ans$ver<-gotVer;
 ans$handlers<-list();
 class(ans)<-"RethinkDB_connection";
 ans
}

#' Check if connection is opened
#'
#' Check whether a given connection is opened.
#' Closed connections cannot be used, and will throw errors on such an attempt; their associated callbacks and/or sync cursor are dead and won't fire/produce any more data.
#'
#' @param x Connection to to check.
#' @return \code{TRUE} if connection is opened and can be used for queries, \code{FALSE} otherwise.
#' @author Miron B. Kursa
#' @export
isOpened<-function(x){
 stopifnot(inherits(x,"RethinkDB_connection"));
 opened<-try(isOpen(x$con),silent=TRUE);
 if(inherits(opened,"try-error")){
  FALSE;
 }else{
  opened;
 }
}

#' Print RethinkDB connection
#'
#' Prints a RethinkDB connection details, including a number of pending callbacks.
#'
#' @param x Connection to print.
#' @param ... Ignored.
#' @method print RethinkDB_connection
#' @note Never blocks.
#' @export
print.RethinkDB_connection<-function(x,...){
 opened<-ifelse(isOpened(x),"Opened","Lost");
 cat(sprintf("\n %s connection to RethinkDB @ %s:%s.\n",
  opened,x$host,x$port));
 if(length(x$handlers)>0){
  cat(sprintf(" %s handlers await draining.\n",length(x$handlers)))
 }
 cat("\n");
}

#' Close RethinkDB connection
#'
#' Closes connection and stops all associated callbacks and/or sync cursor.
#'
#' @param con Connection to close.
#' @param ... Ignored.
#' @method close RethinkDB_connection
#' @export
close.RethinkDB_connection<-function(con,...){
 #Closing socket
 try(close(con$con),silent=TRUE);
 try(close(con$currentCursor),silent=TRUE);
 #Removing handlers
 con$handlers<-list();
 return(con);
}

START<-1;
CONTINUE<-2;
STOP<-3;
sendQuery<-function(x,type,query,token,options){
 stopifnot(inherits(x,"RethinkDB_connection"));

 if(missing(token))
  token<-as.integer(floor(stats::runif(1)*2^31));

 if(nchar(query)>0){
  if(missing(options)||(length(options)==0)){
   query<-sprintf("[%s,%s]",type,query);
  }else{
   query<-sprintf("[%s,%s,%s]",type,query,rjson::toJSON(options));
  }
 }else{
  query<-sprintf("[%s]",type);
  if(!missing(options)) stop("Can't have options without query.");
 }

 len<-as.integer(nchar(query)+1); #Also \0 at the end!
 writeBin(token,x$con,size=8,endian='little');
 writeBin(len,x$con,size=4,endian='little');
 writeBin(query,x$con,endian='little');

 return(token);
}

fetchResponseRaw<-function(x,syncToken){
 stopifnot(inherits(x,"RethinkDB_connection"));

 token<-readBin(x$con,integer(),size=8);
 if(length(token)==0){
  Sys.sleep(1);
  return(x);
 }
 len<-readBin(x$con,integer(),size=4);
 U<-readBin(x$con,raw(),size=1,n=len);
 stuff<-rjson::fromJSON(readBin(U,character()));

 SUCCESS_PARTIAL<-3;

 if(length(stuff$t)!=1) stop("Invalid response!");

 #Parse response type
 if(stuff$t>15){
  close(x);
  stop(sprintf("Error: %s",stuff$r));
 }

 if(!is.null(stuff$p)){
  message("Saved profile in the connection's lastProfile element.");
  x$lastProfile<-stuff$p;
 }

 token<-token;
 if(!is.null(x$handlers[[sprintf("t%s",token)]])){
  hadCallback<-TRUE;
  isMore<-stuff$t==SUCCESS_PARTIAL;
  cba<-TRUE;
  for(element in stuff$r){
   x$handlers[[sprintf("t%s",token)]](element)->cba;
   if(!identical(cba,TRUE)) break;
  }
  if(isMore){
   if(identical(cba,TRUE)){
    #Request more
    sendQuery(x,CONTINUE,'',token);
   }else{
    #Drop this query
    x$handlers[[sprintf("t%s",token)]]<-NULL;
    sendQuery(x,STOP,'',token);
   }
  }else{
   x$handlers[[sprintf("t%s",token)]]<-NULL;
  }
 }else{
  #No handler...
  if((!missing(syncToken))&&(token==syncToken)){
   #...maybe this is a syncToken?
   return(list(r=stuff$r,cc=stuff$t==SUCCESS_PARTIAL));
  }else{
   #...maybe this is a currentCursor syncToken?
   if((!is.null(x$currentCursor))&&(x$currentCursor$token==token)){
    #Add response to its cache
    x$currentCursor$contents<-c(x$currentCursor$contents,stuff$r);
    x$currentCursor$active<-(stuff$t==SUCCESS_PARTIAL);
   }
  }
  #if not, probably a dangling query stop ack; ignore.
 }
 return(invisible(NULL));
}

hookQuery<-function(x,query,cb,options){
 token<-sendQuery(x,START,query,options=options);
 x$handlers[[sprintf("t%s",token)]]<-cb;
 return(x);
}

syncQuery<-function(x,query,options){
 token<-sendQuery(x,START,query,options=options);
 while(TRUE){
  fetchResponseRaw(x,token)->ans;
  if(is.null(ans)) next; #That was an async response
  break;
 }
 if((length(ans$r)==1)&&(!ans$cc)){
  return(ans$r[[1]]);
 }else{
  if((!is.null(x$currentCursor))&&(x$currentCursor$active)){
   warning("Stopping existing, active sync cursor. Use async API to work with multiple streams!");
   close(x$currentCursor);
  }
  x$currentCursor<-makeCursor(x,ans$r,token,ans$cc);
  return(x$currentCursor);
 }
 return(ans);
}

#' Drain RethinkDB connection
#'
#' Drains a given RethinkDB connection, i.e. pull query responses and both call their associated callbacks (for async queries) and/or filling sync cursor local cache.
#' Draining ends when all async queries end; the function blocks for the entire time this is happening.
#'
#' The async query callback will only fire during \code{drainConnection} or (opportunistically) \code{\link{cursorNext}}; consequently this function must be run to guarantee that installed callbacks will have a chance to fire.
#'
#' @param x Connection to drain.
#' @export
drainConnection<-function(x){
 while(length(x$handlers)>0){
  if(isOpened(x)){
   fetchResponseRaw(x);
  }else{
   close(x); #Also clears x$handlers
  }
 }
 return(invisible(NULL));
}


R/cursor.R
makeCursor<-function(con,contents,token,active){
 new.env()->ans;
 class(ans)<-"RethinkDB_cursor";
 ans$con<-con;
 ans$active<-active; # i.e. makes sense to pull more after contents are used
 ans$contents<-contents;
 ans$token<-token;
 ans
}

#' Pull next object from a cursor
#'
#' Pulls a datum from a given cursor, sending continuation queries when needed.
#'
#' @param cursor Cursor to pull from; a result of \code{r()$...$run(...)}.
#' @param inBatch If set to \code{TRUE}, enables batch mode, i.e., returning the whole local cache (this is usually NOT the whole data available under cursor) rather than a single result.
#' Values other than \code{TRUE} or \code{FALSE} are invalid.
#' @return In a default mode, a list representing the returned response JSON, or \code{NULL} if no data is available.
#' In a batch mode, list of such lists representing the whole cache (which may be empty, corresponding to default mode's \code{NULL}).
#' @note When this function empties local cache, it may ask RethinkDB for more data and hence block.
#' Use \code{\link{isCursorEmpty}} to decide if it makes sense to call \code{cursorNext}.
#' In case you don't need any more answers for the query, close cursor with \code{close} method.
#' @author Miron B. Kursa
#' @seealso \code{\link{isCursorEmpty}},
#' @export
cursorNext<-function(cursor,inBatch=FALSE){
 stopifnot(inherits(cursor,"RethinkDB_cursor"));
 cursor$active<-(cursor$active)&&(!is.null(cursor$con))&&isOpened(cursor$con);
 if((length(cursor$contents)==0)&&(cursor$active)){
  #Empty cache? Query for more!
  sendQuery(cursor$con,CONTINUE,'',cursor$token);
  while(TRUE){
   fetchResponseRaw(cursor$con,cursor$token)->ans;
   if(is.null(ans)) next;
   break;
  }
  cursor$contents<-ans$r;
  cursor$active<-ans$cc;
 }
 #Use cache
 if(length(cursor$contents)>0){
  if(inBatch){
   ans<-cursor$contents;
   cursor$contents<-list();
  }else{
   ans<-cursor$contents[[1]];
   cursor$contents[-1]->cursor$contents;
  }
  return(ans);
 }

 #Nothing to return
 if(inBatch) return(list());
 return(NULL);
}

#' Check if cursor is empty
#'
#' Check whether a given cursor is fully drained and will output no more datum.
#' The function never blocks; also verifies that the underlying connection is useful.
#'
#' @param cursor Cursor to check; a result of \code{r()$...$run(...)}.
#' @return \code{TRUE} if cursor has no more data to return.
#' @note It is possible that \code{\link{cursorNext}} will return \code{NULL} just after \code{isCursorEmpty} returns \code{FALSE}.
#' Changefeeds cursors (made with \code{r()$...$changes()$...}) will never become empty (provided that connection won't become broken).
#' @author Miron B. Kursa
#' @seealso \code{\link{cursorNext}}
#' @export
isCursorEmpty<-function(cursor){
 stopifnot(inherits(cursor,"RethinkDB_cursor"));
 cursor$active<-(cursor$active)&&(!is.null(cursor$con))&&isOpened(cursor$con);
 return((length(cursor$contents)==0)&&(!cursor$active));
}

#' Print cursor
#'
#' Prints a given cursor's status.
#'
#' @param x Cursor to print.
#' @param ... Ignored.
#' @method print RethinkDB_cursor
#' @note Never blocks; also checks whether the underlying connection is alive.
#' @export
print.RethinkDB_cursor<-function(x,...){
 empty<-isCursorEmpty(x);
 if(empty){
  cat("\n Empty RethinkDB cursor.\n")
 }else{
  cat(sprintf("\n Active RethinkDB cursor;\n %s response(s) cached",
   length(x$contents)));
  if(x$active){
   cat(", more to come.\n");
  }else{
   cat(", no more on the server.\n");
  }
 }
 cat("\n");
 return(invisible(x));
}

#' Close cursor
#'
#' Closes a given cursor and stops its associated query.
#' Should be called on the current cursor before a new sync query is invoked on the same connection.
#'
#' @param con Cursor to close.
#' @param ... Ignored.
#' @method close RethinkDB_cursor
#' @export
close.RethinkDB_cursor<-function(con,...){
 x<-con;
 x$active<-(x$active)&&(!is.null(x$con))&&isOpened(x$con);
 if(x$active)
  sendQuery(x$con,STOP,'',x$token);
 x$con<-NULL; #GC may use it to kill connection
 x$active<-FALSE;
 x$contents<-list();
 return(x);
}


R/reql.R
commCodes<-c(makeArray=2,makeObj=3,var=10,javascript=11,uuid=169,http=153,error=12,implicitVar=13,db=14,table=15,get=16,getAll=78,eq=17,ne=18,lt=19,le=20,gt=21,ge=22,not=23,add=24,sub=25,mul=26,div=27,mod=28,floor=183,ceil=184,round=185,append=29,prepend=80,difference=95,setInsert=88,setIntersection=89,setUnion=90,setDifference=91,slice=30,skip=70,limit=71,offsetsOf=87,contains=93,getField=31,keys=94,values=186,object=143,hasFields=32,withFields=96,pluck=33,without=34,merge=35,betweenDeprecated=36,between=182,reduce=37,map=38,fold=187,filter=39,concatMap=40,orderBy=41,distinct=42,count=43,isEmpty=86,union=44,nth=45,bracket=170,innerJoin=48,outerJoin=49,eqJoin=50,zip=72,range=173,insertAt=82,deleteAt=83,changeAt=84,spliceAt=85,coerceTo=51,typeOf=52,update=53,delete=54,replace=55,insert=56,dbCreate=57,dbDrop=58,dbList=59,tableCreate=60,tableDrop=61,tableList=62,config=174,status=175,wait=177,reconfigure=176,rebalance=179,sync=138,grant=188,indexCreate=75,indexDrop=76,indexList=77,indexStatus=139,indexWait=140,indexRename=156,setWriteHook=189,getWriteHook=190,funcall=64,branch=65,or=66,and=67,forEach=68,func=69,asc=73,desc=74,info=79,match=97,upcase=141,downcase=142,sample=81,default=92,json=98,toJsonString=172,iso8601=99,toIso8601=100,epochTime=101,toEpochTime=102,now=103,inTimezone=104,during=105,date=106,timeOfDay=126,timezone=127,year=128,month=129,day=130,dayOfWeek=131,dayOfYear=132,hours=133,minutes=134,seconds=135,time=136,monday=107,tuesday=108,wednesday=109,thursday=110,friday=111,saturday=112,sunday=113,january=114,february=115,march=116,april=117,may=118,june=119,july=120,august=121,september=122,october=123,november=124,december=125,literal=137,group=144,sum=145,avg=146,min=147,max=148,split=149,ungroup=150,random=151,changes=152,args=154,binary=155,geojson=157,toGeojson=158,point=159,line=160,polygon=161,distance=162,intersects=163,includes=164,circle=165,getIntersecting=166,fill=167,getNearest=168,polygonSub=171,minval=180,maxval=181);

countArgs<-function(f)
 length(as.list(args(f)))-1;

setReqlClass<-function(x){
 class(x)<-"reql"; x
}

lockBaseQuery<-function(x){
 x$baseQuery<-x$query;
 return(x)
}

isReql<-function(x)
 inherits(x,"reql");

enc<-function(com)
 return(as.numeric(commCodes[com]))

coerceDatum<-function(datum){
 if(inherits(datum,"call")){
  stop("Datum is a call!")
 }
 if(inherits(datum,"pairlist")){
  stop("Datum is a pairlist!")
 }
 ## Do not process what is already done
 if(isReql(datum)){
  if(is.environment(datum)){
   #Here, we collapse a reql tree into a query;
   ans<-setReqlClass(datum$query);
   # but this means reql tree is useless, so we can trim it to root,
   # in order to make it re-usable (useful when it is r()$var(..))
   datum$query<-datum$baseQuery;
   return(ans);

  }
  return(datum);
 }

 ## Fix JSON-unfriendly stuff
 #Drop functions
 if(is.function(datum))
 stop("Functions can only exist as direct term arguments.")
 #Environments -> lists
 if(is.environment(datum)) datum<-as.list(datum);
 #Factors -> character
 if(is.factor(datum)) datum<-stats::setNames(as.character(datum),names(datum));
 #Named stuff -> lists
 if(!is.null(names(datum))&&!is.list(datum)) datum<-as.list(datum);

 ## Drill down lists
 if(is.list(datum))
  datum<-lapply(datum,coerceDatum);

 ## Return time
 #Named lists -> those shall be objects, just return
 if(is.list(datum)&&!is.null(names(datum)))
  return(setReqlClass(datum));
 #Un-named stuff with length>1 or a list -> makeArray of elements
 if(length(datum)>1||is.list(datum))
  return(setReqlClass(list(enc("makeArray"),datum)));
 #Single-element vector... scalar
 return(setReqlClass(datum));
}

incorporateTerm<-function(argRaw,id,Q){
 #Convert NULL argRaw meaning argless-function into list()
 if(is.null(argRaw)) argRaw<-list()

 ## Coerce arguments ##
 if(length(argRaw)>0) for(e in 1:length(argRaw)){
  #Evalute it now, not before
  thisArg<-eval(argRaw[[e]],envir=parent.frame(2),enclos=parent.frame(3));
  #Functions got executed with reql arguments
  if(inherits(thisArg,"function")){
   countArgs(thisArg)->arity;
   if(arity==0)
    stop("Anonymous functions without parameters are not supported.");
   internalArgs<-lapply(1:arity,function(e) lockBaseQuery(r()$var(e)));
   body<-do.call(thisArg,internalArgs);
   if(is.environment(body))
    body<-setReqlClass(body$query);
   argRaw[[e]]<-setReqlClass(list(
    enc("func"),
    list(
     list(enc("makeArray"),as.list(1:arity)),
      body
     )
    ));
  }else{
   #Static objects get coerced and makeArray-ied when needed
   argRaw[[e]]<-coerceDatum(thisArg);
  }
 }

 #Following operations will finally kill argRaw pairlist structure
 if(!is.null(names(argRaw))){
  #Some options must be extracted
  argRaw[nchar(names(argRaw))>0]->argOpts;
  argRaw[nchar(names(argRaw))==0]->argArgs;
  names(argArgs)<-NULL;
 }else{
  #No options at all
  argArgs<-argRaw;
  argOpts<-NULL;
 }

 if(!is.null(Q$query)) Q$query<-list(Q$query);
 argArgs<-c(Q$query,argArgs);

 if(is.null(argOpts)){
  Q$query<-list(enc(id),argArgs);
 }else{
  Q$query<-list(enc(id),argArgs,argOpts);
 }
}

funGen<-function(id,Q){
 id<-force(id);
 function(...){
  #This will be used to manipulate with R eval order
  argRaw<-match.call(expand.dots=FALSE)$...;
  incorporateTerm(argRaw,id,Q);
  Q;
 }
}


#' @rdname r
#' @title ReQL root
#' @description Creates ReQL root for building a query.
#' @param db DB name; this is optional, and is just a syntax sugar for \code{r()$db(db)}.
#' @param table Table name; this is optional, requires db to be given, and is just a syntax sugar for \code{r()$db(db)$table(table)}
#' @return ReQL root; use \code{$} (or \code{[[]]}) to chain query terms (like \code{r()$db("test")$table("test")}).
#' In general, anonymous attributes are passed as attributes while named as term options.
#' In context of term arguments, named lists are treated as JSON objects (following \code{rjson} package heuristics), unnamed lists and simple vectors as JSON arrays; classes and attributes are ignored.
#' Term options should be called in the snake case form (for instance \code{return_changes} not \code{returnChanges}), as documented for the original Python driver.
#' To finalise, use \code{$run} or \code{$runAsync}.
#' For a comprehensive description of all terms, see RethinkDB API reference; here we give an overview of some:
#' \item{run(connection,...)}{Evaluate the query; the function will block until first response from RethinkDB to this query will be received.
#' May return cursor, an object representing a stream of data on which \code{\link{cursorNext}} and \code{\link{cursorToList}} can be used to extract actual information.
#' \code{...} may be used to specify run options, like \code{profile}, \code{durability} or \code{read_mode}.}
#' \item{runAsync(connection,callback,...)}{Evaluate the query; for each datum received \code{x}, run \code{callback(x)}.
#' Callback should return \code{TRUE} to be re-evaluated on proceeding data; any other response will cause the query to be dropped immediately.
#' This function returns immediately; to ask R to start evaluating async queries, run \code{\link{drainConnection}}.
#' Note that callbacks can be also called while \code{$run()} blocks waiting for other query to execute.}
#' \item{bracket(...)}{Implementation of the JavaScript \code{(...)} and Python \code{[...]} operation.}
#' \item{funcall(function,atts)}{Implementation of the JavaScript \code{.do()}; note that the order of arguments is different.}
#' @note ReQL is implemented as an environment, thus is mutable unlike most R objects.
#' To this end, you can use variables for chaining like this \code{r()->query;} \code{query$db("a");} \code{query$table("b")}; but consequently you can't use variables to make a re-usable stub, i.e., this is invalid: \code{r()->query;} \code{query$db("a")$table("aa")$run(...)} \code{query$db("b")$table("bb")$run(...);}.
#'
#' If you get "trying to apply non-function" error, you likely have misspelled term name or trying to use a non-existent one.
#'
#' To view raw AST (at any depth), use \code{$query}.
#' @author Miron B. Kursa
#' @examples \dontrun{
#' #Connect to the RethinkDB instance
#' cn<-openConnection()
#'
#' #Get document count in some_db's some_table
#' r()$db("some_db")$table("some_table")$count()$run(cn)
#' #...same can be done shorter
#' r("some_db","some_table")$count()$run(cn)
#'
#' #Fetch 5 random docs from some_db's some_table...
#' r("some_db","some_table")$sample(5)$run(cn)->cursor
#' #...and present as a list`
#' cursorToList(cursor)
#'
#' #Insert an element
#' r("some_db","some_table")$insert(list(id="new",a=1:10,b=list(c=1,d=2)))$run(cn)
#'
#' #Close connection
#' close(cn)
#' }
#' @export
r<-function(db,table){
 setReqlClass(new.env())->Q;
 Q$query<-NULL;
 Q$baseQuery<-NULL;

 #Populating reql composing environment with functions
 for(com in names(commCodes))
  Q[[com]]<-funGen(com,Q);
 Q$row<-function(...)
  stop("Implicit var (row) is not supported by rethinker; use an actual anonymous function.")
 Q$run<-function(connection,...)
  return(syncQuery(connection,rjson::toJSON(Q$query),list(...)))
 Q$runAsync<-function(connection,cb,...)
  return(hookQuery(connection,rjson::toJSON(Q$query),cb,list(...)));
 Q$expr<-function(x){
  if(!is.null(Q$query))
   stop("$expr only makes sense as a first term.");
  Q$query<-coerceDatum(x);
  Q
 }

 #Applying db and table
 if(!missing(db)){
  Q<-Q$db(force(db));
  if(!missing(table))
   Q<-Q$table(force(table));
 }else{
  if(!missing(table))
   stop("You can't give table without specifying db.");
 }

 #Done
 Q
}
